File: week6_dynamic_programming2/1_maximum_amount_of_gold/knapsack.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using std::vector;
using namespace std;
int optimal_weight(int W, const vector<int> &w) {
  int n=w.size();
  vector<vector<int> >value(n+1,vector<int>(W+1));
  for(int i=0;i<=n;i++)
  {
    value[i][0]=0;
  }
  for(int j=0;j<=W;j++)
  {
    value[0][j]=0;
  }
  for(int i=1;i<=n;i++)
  {
    for(int j=1;j<=W;j++)
    {
      if(w[i-1]>j)
      {
        value[i][j]=value[i-1][j];
      }
      else
      {
        value[i][j]=std::max(w[i-1]+value[i-1][j-w[i-1]],value[i-1][j]);
      }
      
    }
  }
  int result = value[n][W];

  return result;

}

int main() {
  int n, W;
  std::cin >> W >> n;
  vector<int> w(n);
  for (int i = 0; i < n; i++) {
    std::cin >> w[i];
  }
  std::cout << optimal_weight(W, w) << '\n';
}

File: week6_dynamic_programming2/1_maximum_amount_of_gold/knapsack.py
# Uses python3
import sys

def optimal_weight(W, w):
    n=len(w)
    weights=[[None]*(n+1) for i in range(W+1)]
    for i in range(n+1):
        for j in range(W+1):
            if i==0 or j==0:
                weights[i][j]=0
            elif w[i-1]<=j:
                weights[i][j]=max(w[i-1]+weights[i-1][j-w[i-1]],weights[i-1][j])
            else:
                weights[i][j]=weights[i-1][j]
    return weights[n][W]


if __name__ == '__main__':
    input = sys.stdin.read()
    W, n, *w = list(map(int, input.split()))
    print(optimal_weight(W, w))
